transport-classes-io_uring/src/main/java/io/netty/channel/uring/IoUringBufferRing.java

```diff
@@ -0,0 +1,413 @@
+    private short nextIndex;
+    private boolean hasSpareBuffer;
+    private IoUringIoHandler source;
+
+    IoUringBufferRing(int ringFd, long ioUringBufRingAddr,
+                      short entries, short bufferGroupId,
+                      int chunkSize, IoUringIoHandler ioUringIoHandler
+    ) {
+        this.ioUringBufRingAddr = ioUringBufRingAddr;
+        this.entries = entries;
+        this.bufferGroupId = bufferGroupId;
+        this.ringFd = ringFd;
+        this.userspaceBufferHolder = new ByteBuf[entries];
+        this.nextIndex = 0;
+        this.chunkSize = chunkSize;
+        this.hasSpareBuffer = false;
+        this.source = ioUringIoHandler;
+    }
+
+    public void markReadFail() {
+        hasSpareBuffer = false;
+    }
+
+    public boolean hasSpareBuffer() {
+        return hasSpareBuffer;
+    }
+
+    void recycleBuffer(short bid) {
+        source.submitBeforeIO(new Runnable() {
+            @Override
+            public void run() {
+                addToRing(bid, true);
+            }
+        });
+    }
+
+    void addToRing(short bid, boolean needAdvance) {
+        ByteBuf byteBuf = userspaceBufferHolder[bid];
+        int mask = entries - 1;
+        long tailFieldAddress = ioUringBufRingAddr + Native.IO_URING_BUFFER_RING_TAIL;
+        short oldTail = PlatformDependent.getShort(tailFieldAddress);
+        int ringIndex = oldTail & mask;
+        long ioUringBufAddress = ioUringBufRingAddr + (long) Native.SIZEOF_IOURING_BUF * ringIndex;
+        PlatformDependent.putLong(ioUringBufAddress + Native.IOURING_BUFFER_OFFSETOF_ADDR, byteBuf.memoryAddress());
+        PlatformDependent.putInt(ioUringBufAddress + Native.IOURING_BUFFER_OFFSETOF_LEN, (short) byteBuf.capacity());
+        PlatformDependent.putShort(ioUringBufAddress + Native.IOURING_BUFFER_OFFSETOF_BID, bid);
+        if (needAdvance) {
+            advanceTail(1);
+        }
+    }
+
+    void appendBuffer(ByteBufAllocator byteBufAllocator, int count) {
+        int expectedIndex = nextIndex + count;
+        if (expectedIndex > entries) {
+            throw new IllegalStateException("Buffer ring is full");
+        }
+
+        for (int i = 0; i < count; i++) {
+            ByteBuf buffer = byteBufAllocator.ioBuffer(chunkSize);
+            userspaceBufferHolder[nextIndex] = buffer;
+            addToRing(nextIndex, false);
+            nextIndex++;
+        }
+        advanceTail(count);
+    }
+
+    ByteBuf borrowBuffer(int bid, int maxCap) {
+        ByteBuf byteBuf = userspaceBufferHolder[bid];
+        ByteBuf slice = byteBuf.slice(0, maxCap);
+        return new UserspaceIoUringBuffer(maxCap, (short) bid, slice);
+    }
+
+    private void advanceTail(int count) {
+        long tailFieldAddress = ioUringBufRingAddr + Native.IO_URING_BUFFER_RING_TAIL;
+        short oldTail = PlatformDependent.getShort(tailFieldAddress);
+        short newTail = (short) (oldTail + count);
+        PlatformDependent.putShortVolatile(tailFieldAddress, newTail);
```

Comment:
why volatile?

Reply:
In fact, it should be “store release”. [liburing](https://github.com/axboe/liburing/blob/1a5e9e65b7519b2f82298c9df6130da2099d580a/src/include/liburing.h#L1583)
However, this memory order is not available in the Unsafe class of Java 8.

Reply:
how is not? it is...it's the putOrdered

Reply:
> how is not? it is...it's the putOrdered

The JDK has putOrderedInt and putOrderedLong, but there is no putOrderedShort. Do you have any good ideas?

Reply:
yep, place the right barriers.
If the short is aligned, you can just use storeBarrier + putShort - that is a decent one

Reply:
you gotta change the order, when you "release store" you have:
```
// whatever is stored or loaded here won't "move" past the fence!
StoreStore; LoadStore (i.e. storeFence)
// the store here cannot happen before the fence and the prev load/stores
store value

```